Check http errors not graphql

MichiTha opened the pull request:
- check for HTTP errors in RelayConnectionFetcher Retrys.
- ignore GraphQL errors, because they should be handled by the transformer

MichiTha committed:
added ignoreDataErrors option to RelayConnection fetcher

MichiTha committed:
check for http errors in RelayConnectionFetcher for retrys and not for graphQL errors

sambou commented on src/fetching/RelayConnectionFetcher.js:
```diff
@@ -72,22 +72,17 @@ export default class RelayConnection<T, Data> implements Fetcher<Data[]> {
 
-			if (data.errors && Array.isArray(data.errors)) {
+				return getConnection(data);
+			} catch (error) {
```
I don't think this will catch errors like 429s or 500s, right?

MichiTha replied:
... surely not

sambou replied:
I think it should. What do you think of this:

```javascript
while (retries < this.maxRetries) {
    try {
       const res = await fetch(url, options);
       const data = await res.json();
        // retry on TOO MANY REQUESTS and server errors
        if (res.status === 429 || res.status > 499) {
            retries = retries + 1;
        } else {
            if (data.errors && Array.isArray(data.errors) && handleErrors) {
                handleErrors(data);
            }
            return getConnection(data);
        }
    } catch(error) {
        console.log("RelayConnectionFetcher: Received errors while fetching connection: ", error);
		retries = retries + 1;
    }
}
```

sambou commented on src/fetching/RelayConnectionFetcher.js:
```diff
@@ -48,7 +48,7 @@ export default class RelayConnection<T, Data> implements Fetcher<Data[]> {
 			const data: GraphQlConnection<T> = await this.fetchConnection();
 
-			yield data.edges.map(edge => this.transformResult(edge.node));
+			yield data.edges.map(edge => this.transformResult(edge.node, data.errors));
```
I don't think it makes sense to pass the error to transformResult, this would be called many times when fetching a connection, but it's enough to call it once per page fetch.

MichiTha replied:
but each fetch could contain another graphQL error, that transfromResult could handle

sambou replied:
Yeah, but `transformResult` gets called for each node in the result connection. If you fetch 4 slices of a connection with 10 results each, it would be enough to handle the error 4 times, not 4 * 10 times (which would happen with this solution).

MichiTha replied:
right... so mabye we should pass it to the getConnection function then?

sambou replied:
I don't we should change these functions, I'd prefer a separate optional callback like I mentioned [here](https://github.com/tristramjs/tristram/pull/16#discussion_r218418863)

MichiTha replied:
With this implementation, it would be possible to log messages or throw errors if the response includes GraphQL errors. But i think it should be possible to change the transformed output according to the errors.

sambou replied:
Sounds reasonable. Why not have both? Then we could `transformResults` also based on errors, and separately handle them to e.g. exit early etc.

MichiTha replied:
... or we could add a option that a erlay exit could be triggerd with transformResults.
With that an exit could also be triggerd if some imporant graphQL fields are missing (which wont throw any errors)

sambou replied:
Yup, actually we could do it even without adding another option. Simply throwing in `.map` should exit early.

sambou replied:
Just leave it like it is right now?

MichiTha replied:
atm `transformResults` just gets the node of a connection

sambou replied:
but in your PR you added `data.errors`, check [l69](https://github.com/tristramjs/tristram/pull/16/files#diff-83053532852658168c27b202bb7c36f3R69)

MichiTha replied:
yeah sure, but this conversation is about this line.
So as this conversation still is not resolved, i thougt you meant that this would already be the case whitout the change in this line with this PR 😕

sambou replied:
Oh, sorry for the confusion!

MichiTha committed:
check for http statusCodes

MichiTha commented:
is just added the `res.statusCode` check for retrys and error throwing and mad the error codes for retrys and exit agnostic.

Because there are some GraphQL Backends that don't respond with 200 code even if just a small part of a GraphQL response could not be resolved, it should be parameterized.

MichiTha commented on src/fetching/RelayConnectionFetcher.js:
```diff
@@ -40,6 +42,8 @@ export default class RelayConnection<T, Data> implements Fetcher<Data[]> {
 		this.chunkSize = chunkSize || 100;
 		this.logErrors = logErrors || false;
+		this.retrieCodes = retrieCodes || [ 429, 500 ];
+		this.exitCodes = exitCodes || [ 404 ];
```
here we should discuss the default values

sambou commented on src/fetching/RelayConnectionFetcher.js:
```diff
@@ -72,22 +76,23 @@ export default class RelayConnection<T, Data> implements Fetcher<Data[]> {
-			if (data.errors && Array.isArray(data.errors)) {
+			try {
+				const res = await fetch(url, options);
+				if (retrieCodes.includes(res.statusCode)) {
```
I think you're looking for `status`. See [here](https://developer.mozilla.org/en-US/docs/Web/API/Response/status)

MichiTha committed:
use correct status code

MichiTha committed:
remove package-lock

sambou commented on src/fetching/RelayConnectionFetcher.js:
```diff
@@ -19,9 +19,11 @@ export default class RelayConnection<T, Data> implements Fetcher<Data[]> {
 	query: string;
 	chunkSize: number;
-	transformResult: T => Data;
+	transformResult: (T, ?Array<{ message: string, ...any }>) => Data;
```
Remove `...any`, shouldn't be needed here.

sambou commented on src/fetching/RelayConnectionFetcher.js:
```diff
@@ -19,9 +19,11 @@ export default class RelayConnection<T, Data> implements Fetcher<Data[]> {
+	transformResult: (T, ?Array<{ message: string, ...any }>) => Data;
 	logErrors: boolean;
 	maxRetries: number;
+	retrieCodes: [number];
```
Please rename to `retryCodes`.

MichiTha committed:
removed ...any on error type and added some http codes for error handling

MichiTha committed:
renamed retrieCodes to retryCodes

sambou commented on src/fetching/RelayConnectionFetcher.js:
```diff
@@ -72,22 +90,24 @@ export default class RelayConnection<T, Data> implements Fetcher<Data[]> {
+					throw new Error(`server response error ${status}`);
+				} else {
+					const data = await res.json();
+					return getConnection(data);
```
here you also need to somehow return the errors if you want to use them in `transformResult`

MichiTha replied:
the graphQL errors are already inside of `data` (`data.errors`).
So you meant http errors?

sambou replied:
No, i mean that it would make sense to return sth. like
```javascript
{...getConnection(data), errors: data.errors}
```

Otherwise, all clients need to remember to do that in their `getConnection` implementations

MichiTha committed:
fixed types, return errors from fetchConnection

MichiTha committed:
adjusted tests for http error handling

MichiTha commented:
i just adjusted the tests for HTTP error handling and changed the return values for `fetchConnection` for `getData` generator.

so i hope that everything should be fine now for new version release

MichiTha committed:
more tests

MichiTha committed:
bump version

MichiTha committed:
bump minor version
